src/Processors/Merges/Algorithms/Graphite.cpp

```diff
@@ -0,0 +1,471 @@
+    if (a.function != nullptr)
+        stream << ", function = " << a.function->getName();
+    if (!a.retentions.empty())
+    {
+        stream << ",\n  retentions = {\n";
+        for (size_t i = 0; i < a.retentions.size(); i++)
+        {
+            stream << "    { " << a.retentions[i].age << ", " << a.retentions[i].precision << " }";
+            if (i < a.retentions.size() - 1)
+                stream << ",";
+            stream << "\n";
+        }
+        stream << "  }\n";
+    }
+    else
+        stream << " ";
+
+    stream << "}";
+    return stream;
+}
+
+std::string buildTaggedRegex(std::string regexp_str)
+{
+    boost::erase_all(regexp_str, " "); /* cleanup spaces */
+    size_t name_pos = regexp_str.find(';');
+    if (name_pos != regexp_str.npos)
+    {
+        /* tag in format
+         *
+         * tag1=value1 ; tag2=VALUE2_REGEX ; tag3=value3
+         * or
+         * name ; tag1=value1 ; tag2=VALUE2_REGEX ; tag3=value3
+         * or for one tag add ';' to the end
+         * tag1=value1 ;
+         */
+
+        std::vector<std::string> tags;
+
+        boost::split(tags, regexp_str, boost::is_any_of(";"));
+        /* remove empthy elements */
+        using namespace std::string_literals;
+        tags.erase(std::remove(tags.begin(), tags.end(), ""s), tags.end());
+        if (tags[0].find('=') == tags[0].npos)
+        {
+            if (tags.size() == 1) /* only name */
+                return tags[0] + "\\?";
+            /* start with name value */
+            regexp_str = tags[0] + "\\?(.*&)?";
+            tags.erase(std::begin(tags));
+        }
+        else
+            regexp_str = "[\\?&]";
+
+        std::sort(std::begin(tags), std::end(tags)); /* sorted tag keys */
+        regexp_str += boost::algorithm::join(tags, "&(.*&)?");
+        regexp_str += "(&.*)?$"; /* close regex */
+        return regexp_str;
+    }
+    return regexp_str;
+}
+
+/** Read the settings for Graphite rollup from config.
+  * Example
+  *
+  * <graphite_rollup>
+  *     <path_column_name>Path</path_column_name>
+  *     <pattern>
+  *         <regexp>click_cost</regexp>
+  *         <function>any</function>
+  *         <retention>
+  *             <age>0</age>
+  *             <precision>3600</precision>
+  *         </retention>
+  *         <retention>
+  *             <age>86400</age>
+  *             <precision>60</precision>
+  *         </retention>
+  *     </pattern>
+  *     <default>
+  *         <function>max</function>
+  *         <retention>
+  *             <age>0</age>
+  *             <precision>60</precision>
+  *         </retention>
+  *         <retention>
+  *             <age>3600</age>
+  *             <precision>300</precision>
+  *         </retention>
+  *         <retention>
+  *             <age>86400</age>
+  *             <precision>3600</precision>
+  *         </retention>
+  *     </default>
+  * </graphite_rollup>
+  */
+static RuleType
```

Comment:
I find it a little bit confusing. I'd expect append, maybe, returning the pattern (or pointer to it, to save some mem). But append returning a type looks a bit strange to me

Reply:
RuleType is enum. append return a type, and result is checked if a generic rule (RuleTypeAll) or not ? May be change returned result to bool ? But how to transparent interpeter them if false - rule don't added or not RuleTypeAll ?